source/common/quic/envoy_quic_proof_source.cc

```diff
@@ -111,23 +73,19 @@ EnvoyQuicProofSource::getTlsCertConfigAndFilterChain(const quic::QuicSocketAddre
   if (filter_chain == nullptr) {
     listener_stats_.no_filter_chain_match_.inc();
     ENVOY_LOG(warn, "No matching filter chain found for handshake.");
-    return {absl::nullopt, absl::nullopt};
+    return {};
   }
   ENVOY_LOG(trace, "Got a matching cert chain {}", filter_chain->name());
 
   auto& transport_socket_factory =
       dynamic_cast<const QuicServerTransportSocketFactory&>(filter_chain->transportSocketFactory());
```

Comment:
If you split this part out, you should be able to use a bool in QuicServerTransportSocketFactory to guard the new behaviors both in this class and in QuicServerTransportSocketFactory.